astropy/nddata/nduncertainty.py

```diff
@@ -30,155 +84,187 @@ class MissingDataAssociationException(Exception):
+         another instance or subclass of `NDUncertainty`). For subclasses that
+         cannot propagate with arbitary other uncertainties they should check
+         that the other uncertainty has the same class as they are *or* convert
+         it to such a class. (Handling units should be part of the
+         ``_propagate_*`` methods and *NOT* done in there).
+       - The :meth:`NDUncertainty.propagate` method should
+         only be overriden if one wants to allow other operations than
+         ``addition``, ``subtraction``, ``multiplication`` or ``division``.
+         This function is the common entry point of any arithmetic computation
+         that tries to propagate uncertainties and is responsible for calling
+         the appropriate ``_propagate_*`` method and then returns another
+         instance of the same class with the results uncertainty.
+
+    5. `NDUncertainty` and it's subclasses can only be used for uncertainty
+       propagation if their ``parent_nddata`` attribute is a reference to the
+       `NDData` object. This is needed since many kinds of propagation need the
+       actual data besides the uncertainty.
+
+    6. If a ``unit`` is implicit (``data`` had a unit) or explicitly passed
+       to the ``__init__`` one should be sure that this unit is identical or
+       convertable to the unit of the parent. Otherwise uncertainty propagation
+       should fail.
     """
 
-    __metaclass__ = abc.ABCMeta
-
-    # Indicates whether the class supports the propagation of correlated
-    # uncertainties
-    supports_correlated = False
+    def __init__(self, array=None, unit=None, copy=True):
+        if isinstance(array, NDUncertainty):
+            # Given an NDUncertainty class or subclass check that the type
+            # is the same.
+            if array.uncertainty_type != self.uncertainty_type:
+                raise IncompatibleUncertaintiesException
+            # Check if two units are given and take the explicit one then.
+            if (unit is not None and array.unit is not None and
+                    unit != array.unit):
+                # TODO : Clarify it (see NDData.init for same problem)?
+                log.info("Overwriting Uncertainty's current "
+                         "unit with specified unit")
+            elif array.unit is not None:
+                unit = array.unit
+            array = array.array
 
-    @property
-    def parent_nddata(self):
-        if self._parent_nddata is None:
-            message = "uncertainty is not associated with an NDData object"
-            raise MissingDataAssociationException(message)
+        elif isinstance(array, Quantity):
+            # Check if two units are given and take the explicit one then.
+            if (unit is not None and array.unit is not None and
+                    unit != array.unit):
+                # TODO : Clarify it (see NDData.init for same problem)?
```

Comment:
You silently accepted it in #4270 (I think there is still the TODO). That was if given two units (one from data, one explicitly) should raise an error or raise a warning.